### PR TITLE
Optimistically create policy expense chats for new members of workspaces

### PR DESCRIPTION
**Pull request by robertKozik**
### Details
Add creating optimistic policy expense chats while adding new members to the workspace, so created chats would pop up while user is offline. 

### Fixed Issues

$ https://github.com/Expensify/App/issues/14594 
PROPOSAL: https://github.com/Expensify/App/issues/14594#issuecomment-1466995775

### Tests

Same as QA steps

- [X] Verify that no errors appear in the JS console

### Offline tests

Same as the QA steps. Workspace chats for each member should appear optimistically
### QA Steps

1. Log in with account that has access to the `policy-expense-chat` beta
2. Go offline by navigating to [profile icon] > Preferences > and enable `Force offline`
3. Create a new workspace.
4. Invite some people to the workspace. Verify that workspace chats for each member appear optimistically.
5. Go online by navigating to [profile icon] > Preferences > disable `Force online`. Verify that workspace chats created for each members are updated accordingly

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/61146594/228816124-18a77c59-8dbc-4901-99a3-bcd603266bc6.mov
</details>

<details>

<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/61146594/228822830-49b30a11-21f3-47f9-8cd1-b5c24f78bb42.MP4
</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/61146594/228822050-f2aa1bc7-a585-434b-99f5-f671747aee98.mov
</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/61146594/228823947-bcddeb92-6c9d-4a3d-8d5f-6dd697c1f159.mov
</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/61146594/228819146-10c9783c-5c18-4052-916e-5bed65dd94aa.mov
</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/61146594/228858213-4561064b-9910-48b7-9121-19b111114467.mov
</details>

### CODE REVIEW

**Comment by RobertKozik** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by bondydaa** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

File: src/libs/actions/Policy.js [hunk 1]
```diff
@@ -222,17 +223,91 @@ function removeMembers(members, policyID) {
...
  * @param {String} policyID
+ * @param {Array<String>} betas
  */
-function addMembersToWorkspace(memberLogins, welcomeNote, policyID) {
+function addMembersToWorkspace(memberLogins, welcomeNote, policyID, betas) {
     const membersListKey = `${ONYXKEYS.COLLECTION.POLICY_MEMBER_LIST}${policyID}`;
     const logins = _.map(memberLogins, memberLogin => OptionsListUtils.addSMSDomainIfPhoneNumber(memberLogin));
 
+    // create onyx data for policy expense chats for each new member
+    const membersChats = createPolicyExpenseChats(policyID, logins, betas);
```
Comment (bondydaa) [inline src/libs/actions/Policy.js:309]: NAB maybe a better variable name would be like `membersChatData`

**Review by bondydaa — DISMISSED** [review]
Just one small thing from me that you can change or leave as is, I don't have a strong opinion. 

Tested on desktop and 👍 :chef-kiss:

**Comment by bondydaa** [comment]
all you @0xmiroslav let me know if you need help with the checklist.

**Comment by RobertKozik** [comment]
@0xmiroslav rebased, so should be up to date

**Review by bondydaa — DISMISSED** [review]

**Comment by RobertKozik** [comment]
> BUG: multiple workspace chats are created for the same user

@0xmiroslav I cannot reproduce bug you discovered, I'm attaching the video - tried to reproduce what I was on yours as much as possible

https://user-images.githubusercontent.com/61146594/230592535-73a9a244-832f-41d1-bb00-c262ee3014e0.mov

@bondydaa I remember the similar problem while implementing this issue - returned report ID is not matched with optimistically created one, thus another report is shown in LHN

File: src/libs/actions/Policy.js [hunk 2]
```diff
@@ -222,17 +223,91 @@ function removeMembers(members, policyID) {
...
+        workspaceMembersChats.onyxOptimisticData.push({
+            onyxMethod: CONST.ONYX.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${optimisticReport.reportID}`,
+            value: {[optimisticCreatedAction.reportActionID]: optimisticCreatedAction},
+        });
+        workspaceMembersChats.onyxSuccessData.push({
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${optimisticReport.reportID}`,
+            value: {
+                isLoadingReportActions: false,
```
[the remark this answers is not in the document]
Comment (RobertKozik) [reply src/libs/actions/Policy.js:277]: I was modelling that on creating report inside `Report.js` file in function `openReport` (393:400 lines)

**Comment by bondydaa** [comment]
> should we also cross-out workspace chat when remove user from workspace offline?

Hmm I didn't considered that scenario initially. 

I think let's try to keep the scope of this PR small and only focused on the `AddMembersToWorkspace` API method since I'm assuming the removing of a user is a different API command? We can spin up a new issue to ensure the offline behavior for removing a workspace member is handled properly.

File: src/libs/actions/Policy.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bondydaa) [reply src/libs/actions/Policy.js:277]: Looks like `false` is the default here https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/pages/home/ReportScreen.js#L96

and it's used here https://github.com/Expensify/App/blob/64f4a7814e87bd7c4d085776abf5cb8a67b9e8bc/src/pages/home/ReportScreen.js#L228

as well as here
```
src/pages/home/report/ReportActionsList.js
180:                        if (this.props.report.isLoadingReportActions && lastReportAction.actionName !== CONST.REPORT.ACTIONS.TYPE.CREATED) {

src/pages/home/report/ReportActionsView.js
134:        if (nextProps.report.isLoadingReportActions !== this.props.report.isLoadingReportActions) {
```

Seems like it can probably be removed.

**Comment by bondydaa** [comment]
I didn't see more than 1 when I tested yesterday on desktop but let me try again. Did you only add 1 member while offline? Or multiple? I think I tested adding 2 users but not just a single user so I'll try both.

**Comment by RobertKozik** [comment]
I've removed isLoadingReportActions key and add failure onyx data as requested in review.

**Comment by bondydaa** [comment]
Hmm interesting, I'll take a look at logs to see if I can confirm the following:

What I suspect is happening is that onyx is just "out of date" with the removed user and that's why you're able to create a new chat successfully which then shows 2 of them since the first one for the user is technically archived which is why I think it's more related to `DeleteMembersFromWorkspace`. 

The behavior is only showing up now in this PR since previously the policy expense chats couldn't be added optimistically anyways.

**Comment by RobertKozik** [comment]
I checked this more in depth. In my opinion this bug consists of two separate problems:

1. When user is offline and tries to remove optimistically workspace member - created chat is not being removed
2. User is invited and chat is created successfully online, later removed - in this case when member is going to be re-added while app is offline, optimistically created chat will be created, but passed `reportID` will be ignored by backend. Returned chat will be the same one as when user was being added for the first time.

The first one is more related to `DeleteMembersFromWorkspace` as @bondydaa stated. But the second one should be resolved in this PR.
I came up with the solution: after removing the user chat remains as archived. We will look for this report and just change it's status to open (optimistically) as API will return it's ID as new chat. After coming back online, everything should work as intended.

**Review by bondydaa — CHANGES_REQUESTED** [review]
One of my PRs has been merged and deployed for the "backend" db servers. I've published my PR for review for the API changes https://github.com/Expensify/Web-Expensify/pull/36993

Here is what I've updated the parameter to
image

Don't change it just yet since a reviewer may ask me to change it so I'll update if that happens.

**Comment by RobertKozik** [comment]
Hi @bondydaa what is the status of your PRs? Can we proceed with this one?

File: src/libs/actions/Policy.js [hunk 3]
```diff
@@ -275,6 +381,7 @@ function addMembersToWorkspace(memberLogins, welcomeNote, policyID) {
         // Escape HTML special chars to enable them to appear in the invite email
         welcomeNote: _.escape(welcomeNote),
         policyID,
+        reportIDs: JSON.stringify(membersChats.reportIDs),
```
[the remark this answers is not in the document]
Comment (bondydaa) [reply src/libs/actions/Policy.js:384]: we left this as `reportCreationData` in the API so feel free to change it now and test.

File: src/libs/actions/Policy.js [hunk 4]
```diff
@@ -269,7 +269,11 @@ function createPolicyExpenseChats(policyID, members, betas) {
         );
         const optimisticCreatedAction = ReportUtils.buildOptimisticCreatedReportAction(optimisticReport.ownerEmail);
 
-        workspaceMembersChats.optimisticReportIDs[login] = optimisticReport.reportID;
+        workspaceMembersChats.reportIDs[login] = {
```
Comment (bondydaa) [inline src/libs/actions/Policy.js:272]: I think you missed a spot when renaming here. Getting an error while testing this 
image
which is linking to this line 
image

```suggestion
        workspaceMembersChats.reportCreationData[login] = {
```

File: src/libs/actions/Policy.js [hunk 5]
```diff
@@ -375,7 +381,7 @@ function addMembersToWorkspace(memberLogins, welcomeNote, policyID, betas) {
         // Escape HTML special chars to enable them to appear in the invite email
         welcomeNote: _.escape(welcomeNote),
         policyID,
-        optimisticReportIDs: JSON.stringify(membersChats.optimisticReportIDs),
+        reportCreationData: JSON.stringify(membersChats.reportCreationData),
```
Comment (bondydaa) [inline src/libs/actions/Policy.js:384]: crap just realizing I might have screwed up the WAF rule for this 🤦 going to see about testing real quick.

File: src/libs/actions/Policy.js (hunk 4 above)
Comment (bondydaa) [reply src/libs/actions/Policy.js:272]: yep confirmed this change gets rid of the errors

File: src/libs/actions/Policy.js (hunk 5 above)
Comment (bondydaa) [reply src/libs/actions/Policy.js:384]: and yes confirmed I did this wrong, let me fix it.

**Review by bondydaa — CHANGES_REQUESTED** [review]

**Comment by bondydaa** [comment]
I know you can't access this link but got a PR up fixing the WAF rule here https://github.com/Expensify/Web-Expensify/pull/37082

I'll get it to staging as soon as it's merged via Cherry Pick (CP).

File: src/libs/actions/Policy.js (hunk 4 above)
Comment (RobertKozik) [reply src/libs/actions/Policy.js]: Sorry for that, commiting changes 🚀

**Review by bondydaa — DISMISSED** [review]

**Comment by bondydaa** [comment]
@0xmiroslav can you handle the the reviewer checklist?

**Comment by RobertKozik** [comment]
I think this one is more related to deleting the member, thus should be considered in: https://github.com/Expensify/App/issues/17137.

File: src/libs/ReportUtils.js [hunk 6]
```diff
@@ -1515,6 +1515,25 @@ function getChatByParticipants(newParticipantList) {
...
 }
 
+/**
+ * Attempts to find a report in onyx with the provided list of participants in given policy
+ * @param {Array} newParticipantList
+ * @param {String} policyID
+ * @returns {object|undefined}
+ */
+function getChatByParticipantsAndPolicy(newParticipantList, policyID) {
+    newParticipantList.sort();
```
[the remark this answers is not in the document]
Comment (RobertKozik) [reply src/libs/ReportUtils.js:1525]: I was modelling this function on the one above it: `getChatByParticipants`.  But you're right, it looks irrelevant

**Comment by bondydaa** [comment]
yeah I agree probably most likely related to deleting the member which we know we need to improve in that other issue.

**Comment by RobertKozik** [comment]
Thats very strange, like info about that second WS changed 😖 
I've locally rebased this branch to be up to date, and I cannot reproduce what you have encountered

https://user-images.githubusercontent.com/61146594/234125054-d93b4855-9593-43df-b924-10ce3dddb9cf.mp4

**Comment by RobertKozik** [comment]
@0xmiroslav I'll be glad If you could confirm me that after merging with main to be up-to-date you can still encounter the bug. Im not sure whether it fixes it, or I've just made mistake in the reproduction

**Comment by RobertKozik** [comment]
Merge done

**Comment by RobertKozik** [comment]
I checked it on web and try to copy tour path, and it works. I'll check more in depth later how it's possible that it's not creating for you.

https://user-images.githubusercontent.com/61146594/234712099-e74d28e3-3bff-47bd-adee-ca81fe225933.mov

**Comment by bondydaa** [comment]
> I think there's serious bug in workspace name offline change logic (maybe not related to PR)

Yep there is and I believe it's related to the "replay problems" we've been having. nothing to do with this PR.

**Comment by bondydaa** [comment]
I'm not able to reproduce what @0xmiroslav showed either. 

Can you double check you were on this branch and try again?

**Comment by bondydaa** [comment]
yeah I've tried multiple ways and can't reproduce but let me make sure my local environment is the same as what the Staging servers have just to be sure as well.

**Comment by bondydaa** [comment]
hmm nope still nodda locally for me, let me switch my new dot to staging as well and see if that does it since the workspace name change offline still reverted back.

**Comment by bondydaa** [comment]
yeah even against staging I'm still seeing the optimistic Expense Policy Chats being created. 

https://user-images.githubusercontent.com/4073354/234720691-e59085b3-8acb-47b4-b91e-3d62e642a851.mp4

**Comment by bondydaa** [comment]
okay kicked off a build for this, workflow is here https://github.com/Expensify/App/actions/runs/4814233404. I guess a QR code should show up in like an hour or so?

**Comment by robertKozik** [comment]
Great catch, fixed 🔧 

https://user-images.githubusercontent.com/61146594/234994030-c689ea83-84cb-42aa-96bd-2a8896864b34.mov

**Review by bondydaa — APPROVED** [review]

**Review by 0xmiroslav — APPROVED** [review]
Looks good, tests well on all possible cases except the known workspace name issue on production backend. This is fixed in staging backend.
Anyway, not a blocker for this PR.

https://user-images.githubusercontent.com/97473779/235013831-7fd4d162-aa82-4093-b017-1e11a0f96861.mov

File: src/libs/actions/Policy.js [hunk 7]
```diff
@@ -227,17 +228,121 @@ function removeMembers(members, policyID) {
...
+        reportCreationData: {},
+    };
+
+    // If the user is not in the beta, we don't want to create any chats
+    if (!Permissions.canUsePolicyExpenseChat(betas)) {
+        return workspaceMembersChats;
+    }
+
+    _.each(members, (login) => {
+        const oldChat = ReportUtils.getChatByParticipantsAndPolicy([sessionEmail, login], policyID);
```
[the remark this answers is not in the document]
Comment (robertKozik) [reply src/libs/actions/Policy.js:253]: `members` variable stores only users which are going to be add to the workspace. But we can get the policy members from **Onyx** store and filter out members to get all admins. The downside is that while being offline there is possibility of not having this info stored, so we would still fallback to the `[newMember, ourMail]` situation like here.

[the remark this answers is not in the document]
Comment (robertKozik) [reply src/libs/actions/Policy.js:253]: It shouldn't be difficult to implement, so If you think it should support multiple admins already, I'm happy to help

[the remark this answers is not in the document]
Comment (bondydaa) [reply src/libs/actions/Policy.js:253]: Hmm I think this is probably better as is for now, if/when we start allowing more than 1 admin on a workspace we can update this code as needed.

**Comment by bondydaa** [comment]
hmm testing this on staging and production right now isn't working anymore 😢

**Comment by bondydaa** [comment]
I think I remember seeing something about us changing the invite member page so my guess it's involved here.

looking at the the logs 
image

**Comment by bondydaa** [comment]
oh think i might have figured it out, testing more, hold on to your butts

**Comment by bondydaa** [comment]
yep confirmed, this https://github.com/Expensify/App/pull/15672 broke it.

**Comment by roryabraham** [comment]
Following up on @bondydaa's comment here, I am seeing that the workspace chats appear to be created optimistically:

image

Then successfully created when coming back online:

image

image

Testing on dev. Let me make sure my Web-E & Auth are up-to-date...

**Comment by roryabraham** [comment]
Oh, maybe I lied and there was an error set on those chats...

**Comment by roryabraham** [comment]
Would love it if we could follow-up with some automated tests for this btw... might have prevented it from breaking

**Comment by Gonals** [comment]
This PR caused [this bug](https://github.com/Expensify/App/issues/42499). A user can be part of a workspace's room (if invited), but not part of the workspace. Inviting those users to the workspace would cause an error, as we'd wrongly identify the room as the Workspace chat.
